Adds Dynamic Button Creation Functionality

Jcobo314 opened the pull request:
This patch allows a user to create buttons with data based on the title of each column in the table. When a button is clicked, the first column of the table will always populate with the current time-stamp of the video being played, then each subsequent column will populate based on data assigned to a button.

Testing Steps:

  1. Run the application and clear all sessions
  2. Created a new session and name it "test"
  3. Click "File" -> "Load video file" then add a video to the application
  4. Click "+" in the "Coding Assistance Buttons" panel
  5. Name this button "button1", assign the hotkey "a",  leave the "Time" field blank and populate each field with data of your choice then click "Create". Do not close the dialog
  6. Repeat step 5 to create a second button with a different name, hotkey, and data. Close the dialog
  7. Press "a" on your keyboard and ensure the table populates with button1's data.
  8. Repeat step 7 with the second button's hotkey. Ensure the data populates and is different from button1's
  9. Click "-" in the "Coding Assistance Buttons" panel
  10. Delete "button1"
  11. Press "a" and ensure no data populates in the table
  12. Close the application

Type: New Functionality

Commit:
dev_dynamic_button_creation Commit #1 (Test)

Commit:
dev_dynamic_button_creation Commit #1 (Test)

Commit:
dev_dynamic_button_creation Commit #2 (Test)

Commit:
dev_dynamic_button_creation Commit #2 (Test)

Briggs-Richardson commented on Controllers/window_controller.py:
```diff
@@ -49,6 +49,7 @@ def __init__(self, window):
         self._window = window
 
         self._manager = Manager()
+        self.encoding_buttons = []
```
I don't think this is being used. If not, then remove.

Briggs-Richardson commented on Controllers/window_controller.py:
```diff
@@ -443,17 +444,20 @@ def create_coding_assistance_button(self):
-        new_button.setObjectName(button_name)
         new_button.setShortcut(QKeySequence(button_hotkey))
-        new_button_definition = ButtonDefinition(new_button, button_name, button_hotkey)
+        self.encoding_buttons.append(new_button)
```
If this serves no purpose, then remove.

Briggs-Richardson commented on View/add_coding_assistance_button_dialog.py:
```diff
@@ -3,12 +3,14 @@
-    def __init__(self):
+    def __init__(self, table):
         """
         Constructor: Initializes the layout of the Add Coding Assistance Button dialog
```
Add a parameters comment, describing what the table parameter is.

Briggs-Richardson commented on View/button_panel.py:
```diff
@@ -74,10 +74,9 @@ def create_coding_assistance_button(self, button_definition):
 
-
     def delete_coding_assistance_button(self, button_name):
         """Deletes a button in the Coding Assistance Panel"""
```
Add a parameters comment, describing what the button_name is.

Briggs-Richardson commented on Application/button_definition.py:
```diff
@@ -6,10 +6,11 @@ class ButtonDefinition:
-    def __init__(self, button, button_id, hotkey):
+    def __init__(self, button, button_id, hotkey, data):
         """
         Constructor - Creates an instance of ButtonDefinition
```
Add a parameters comment, describing what the button, button_id, hotkey, and data are.

Briggs-Richardson commented on Controllers/window_controller.py:
```diff
@@ -463,26 +467,34 @@ def create_coding_assistance_button(self):
-        column = 0
+    def dynamic_button_click(self, button_definition):
+        """
+        Add button data to table when clicked
```
Add a parameters comment, describing what the button_definition is

Briggs-Richardson commented on Controllers/window_controller.py:
```diff
@@ -463,26 +467,34 @@ def create_coding_assistance_button(self):
-        dt_string = date_time.strftime("%d/%m/%Y %H:%M:%S")
-
-        column = 0
+    def dynamic_button_click(self, button_definition):
```
Add a slot annotation above the method: `@Slot(ButtonDefinition)`

Jcobo314 replied:
This isn't a slot function because it's being called within window_controller

Briggs-Richardson reviewed (changes requested):
First CR

Commit:
dev_load_button_dialog Commit #3

Commit:
dev_load_button_dialog Commit #3

Commit:
dev_load_button_dialog Commit #4

Commit:
dev_load_button_dialog Commit #4

Briggs-Richardson committed:
Fixes lowercase session id bug

Before this change, all session ID values written through QSettings were
titleized, using python's title() method. This prevented the user from
creating a session with a lowercase session ID, as it would not equal
its titleized counterpart. This patch removes the title() method,
allowing the session ID to be written as is, and consequently allows for
session IDs of all cases.

This patch also updates the check for session ID uniqueness to be
case-insensitive. This change is to support all platforms, as the
Windows registry uses case-insensitive keys.

Testing Steps:
  1. Run the application and clear all sessions
  2. Create a new session and give it an ID="test"
  3. Write dummy data in the encoding table
  4. Restart the application and verify that the session in the welcome
     screen is labeled "test" and not "Test"
  5. Load the "test" session and verify the dummy data has persisted
  6. Restart the application
  7. Attempt to create a new session called "Test", and verify it is not
     allowed
  8. Repeat steps 1-5, but with a new session ID="Test"

Type: Bug Fix

Briggs-Richardson committed:
Fixes lowercase session id bug

Before this change, all session ID values written through QSettings were
titleized, using python's title() method. This prevented the user from
creating a session with a lowercase session ID, as it would not equal
its titleized counterpart. This patch removes the title() method,
allowing the session ID to be written as is, and consequently allows for
session IDs of all cases.

This patch also updates the check for session ID uniqueness to be
case-insensitive. This change is to support all platforms, as the
Windows registry uses case-insensitive keys.

Testing Steps:
  1. Run the application and clear all sessions
  2. Create a new session and give it an ID="test"
  3. Write dummy data in the encoding table
  4. Restart the application and verify that the session in the welcome
     screen is labeled "test" and not "Test"
  5. Load the "test" session and verify the dummy data has persisted
  6. Restart the application
  7. Attempt to create a new session called "Test", and verify it is not
     allowed
  8. Repeat steps 1-5, but with a new session ID="Test"

Type: Bug Fix

Jcobo314 reviewed (commented):
Testing Step 4 and 7: When creating the new session with the table title set in the session creation page, the title for the table is not updated when the session opens. Could be a future bug fixing task.
Otherwise all testing steps for the new functionality passed

Briggs-Richardson committed:
Adds session mgmt functionality to window toolbar

This patch adds toolbar options, under File, to create and load sessions while
the application is open and running. Upon clicking on either toolbar
option, a new window will display displaying the appropriate session
management window. The user can now load or create a session from within
the application, which if occurs, will reload the window with the
appropriate session data.

Testing Steps:
  1. Run the application and clear all sessions.
  2. Verify that it is possible to navigate to and from the session
     creator page (click "Create New Session", then click "back")
  3. Create a new session "Session A"
  4. Set the table title name to "A" and fill some table cells with
     "A"s.
  5. Restart the application, loading Session A, and verify the "A" data
     persisted.
  6. Restart the application, creating a new session, "Session B"
  7. Set the table title name to "B" and fill some table cells with
     "B"s.
  8. Go to File > Create New Session
  9. Verify the "back" button has been removed.
  10. Close the session creator page.
  11. Go to File > Load Saved Session
  12. Verify all session creator elements are hidden.
  13. Load "Session A" and verify its "A" data persisted.
  14. Load a video file and play it, ensure program functionality works
      as expected.
  15. Go to File > Load Saved Session
  16. Verify "Session B" is now listed
  17. Load "Session B" and verify its "B" data persisted
  18. Repeat step 14
  19. Go to File > Create New Session
  20. Create a new session, "Session C"
  21. Verify a new session has been created in a new window.
  22. Set the table title name to "C" and fill some table cells with
      "C"s
  23. Repeat step 14
  24. Go to File > Load Saved Session
  25. Load "Session B"
  26. Go to File > Load Saved Session
  27. Load "Session C"
  28. Verify the "C" data persisted
  29. Restart the application and verify Sessions A, B, C are listed.
  30. Load Sessions B and C, and verify their data persisted.
  31. Restart the application, and remove "Session B"
  32. Load "Session C"
  33. Go to File > Load Saved Session
  34. Verify that "Session B" is not listed

Type: New Feature

Briggs-Richardson committed:
Adds session mgmt functionality to window toolbar

This patch adds toolbar options, under File, to create and load sessions while
the application is open and running. Upon clicking on either toolbar
option, a new window will display displaying the appropriate session
management window. The user can now load or create a session from within
the application, which if occurs, will reload the window with the
appropriate session data.

Testing Steps:
  1. Run the application and clear all sessions.
  2. Verify that it is possible to navigate to and from the session
     creator page (click "Create New Session", then click "back")
  3. Create a new session "Session A"
  4. Set the table title name to "A" and fill some table cells with
     "A"s.
  5. Restart the application, loading Session A, and verify the "A" data
     persisted.
  6. Restart the application, creating a new session, "Session B"
  7. Set the table title name to "B" and fill some table cells with
     "B"s.
  8. Go to File > Create New Session
  9. Verify the "back" button has been removed.
  10. Close the session creator page.
  11. Go to File > Load Saved Session
  12. Verify all session creator elements are hidden.
  13. Load "Session A" and verify its "A" data persisted.
  14. Load a video file and play it, ensure program functionality works
      as expected.
  15. Go to File > Load Saved Session
  16. Verify "Session B" is now listed
  17. Load "Session B" and verify its "B" data persisted
  18. Repeat step 14
  19. Go to File > Create New Session
  20. Create a new session, "Session C"
  21. Verify a new session has been created in a new window.
  22. Set the table title name to "C" and fill some table cells with
      "C"s
  23. Repeat step 14
  24. Go to File > Load Saved Session
  25. Load "Session B"
  26. Go to File > Load Saved Session
  27. Load "Session C"
  28. Verify the "C" data persisted
  29. Restart the application and verify Sessions A, B, C are listed.
  30. Load Sessions B and C, and verify their data persisted.
  31. Restart the application, and remove "Session B"
  32. Load "Session C"
  33. Go to File > Load Saved Session
  34. Verify that "Session B" is not listed

Type: New Feature

Briggs-Richardson committed:
Updates the encoding button panel layout to 3x3

Before this change, encoding buttons were added in a vertical layout
manager in the button panel. With this change, buttons are added in a
3x3 grid. The buttons are added from top to bottom, left to right.

Testing Steps:
  1. Run the application and create a new session
  2. For each number, x, in the range of 1-9, click on the add encoding
     button, "+", in the encoding panel, and add a button with a label
     of "test button x" and shortcut "x". Verify that the buttons are
     added in a 3x3 grid, from top to bottom, left to right.
  3. Verify that the row and column size did not change throughout step
     2
  4. Add another button of any unique label and shortcut, and verify
     that a ValueError is thrown in the console.
  5. Delete buttons, using their labels, in a randomized order. Verify
     that the buttons are removed from the layout and the layout does
     not contain any empty slots.

Type: New Feature

Briggs-Richardson committed:
Updates the encoding button panel layout to 3x3

Before this change, encoding buttons were added in a vertical layout
manager in the button panel. With this change, buttons are added in a
3x3 grid. The buttons are added from top to bottom, left to right.

Testing Steps:
  1. Run the application and create a new session
  2. For each number, x, in the range of 1-9, click on the add encoding
     button, "+", in the encoding panel, and add a button with a label
     of "test button x" and shortcut "x". Verify that the buttons are
     added in a 3x3 grid, from top to bottom, left to right.
  3. Verify that the row and column size did not change throughout step
     2
  4. Add another button of any unique label and shortcut, and verify
     that a ValueError is thrown in the console.
  5. Delete buttons, using their labels, in a randomized order. Verify
     that the buttons are removed from the layout and the layout does
     not contain any empty slots.

Type: New Feature

Commit:
dev_dynamic_button_creation Commit #5
